torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2437,54 +2437,92 @@ Tensor linalg_qr_backward(const std::vector<torch::autograd::Variable> &grads, c
-  auto singular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
+  auto det_backward_nonsingular = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
+    // Derived from Jacobi's formula for partial derivative, which can be
+    // found at https://en.wikipedia.org/wiki/Jacobi%27s_formula
+    auto det_grad = (unsqueeze_multiple(det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1)).conj();
+    return unsqueeze_multiple(grad, {-1, -2}, self.dim()) * det_grad;
+  };
+
+  auto det_backward_singular_real = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
     Tensor u, sigma, v;
     std::tie(u, sigma, v) = self.svd();
-    auto gsigma = prod_backward(grad.unsqueeze(-1), sigma, det.unsqueeze(-1));
-    return svd_backward({{}, gsigma, {}}, self, true, true, u, sigma, v);
+    auto sigma_grad = prod_backward(grad, sigma, det, -1, false);
+    return svd_backward({{}, sigma_grad, {}}, self, true, true, u, sigma, v);
+  };
+
+  auto det_backward_singular_complex = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
+    Tensor u, sigma, v;
+    std::tie(u, sigma, v) = self.svd();
+    auto u_det = u.det();
+    auto v_det = v.det();
+    auto sigma_prod = sigma.prod(-1);
+    auto sigma_grad = (u_det * v_det).unsqueeze(-1) * prod_backward(grad, sigma, sigma_prod, -1, false);
+    auto u_grad = unsqueeze_multiple(sigma_prod * v_det, {-1, -2}, self.dim()) * det_backward_nonsingular(grad, u, u_det);
+    auto v_grad = unsqueeze_multiple(sigma_prod * u_det, {-1, -2}, self.dim()) * det_backward_nonsingular(grad, v, v_det);
+    return svd_backward({u_grad, sigma_grad, v_grad}, self, true, true, u, sigma, v);
   };
 
-  auto nonsingular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
-    return unsqueeze_multiple(grad * det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+  auto det_backward_singular = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
+    // Formula is derived starting with:
+    //    det(A) = det(u) * prod(sigma) * det(v)
+    //    u, sigma, v = svd(A)
+    //
+    // Using the product rule, we can find that
+    //    d(det(A))/dA = prod(s) * det(v) * d(det(u))/dA
+    //                 + det(u) * det(v) * d(prod(s))/dA
+    //                 + det(u) * prod(s) * d(det(v))/dA
+    //
+    // where "d(x)/dA" indicates the partial derivative of x with
+    // respect to the corresponding element from matrix A.
+
+    if (self.is_complex()) {
+      return det_backward_singular_complex(grad, self, det);
+    } else {
+      // For real matrices, det(u) and det(v) will always be +/-1, and
+      // the previous formula will simplify to
+      //    d(det(A))/dA =  det(u) * det(v) * d(prod(s))/dA
+      //
+      // so we use the simplified implementation in this case.
+      return det_backward_singular_real(grad, self, det);
```

Comment:
`det_backward_singular_real` does not handle when `det(u)/det(v) == -1`

Reply:
Indeed, `det_backward` is just incorrect:
```
In [18]: x                                                                                                                                                                   
Out[18]:                                                                                                                                                                     
tensor([[0.2931, 0.5179, 0.8540],                                                                                                                                            
        [0.6996, 0.3445, 0.8483],                                                                                                                                            
        [0.7505, 0.2611, 0.8554]])                                                                                                                                           
                                                                                                                                                                             
In [19]: u                                                                                                                                                                   
Out[19]:                                                                                                                                                                     
tensor([[-0.5201,  0.8430,  0.1375],                                                                                                                                         
        [-0.6026, -0.2481, -0.7585],                                                                                                                                         
        [-0.6053, -0.4773,  0.6370]])                                                                                                                                        
                                                                                                                                                                             
In [20]: u.det()                                                                                                                                                             
Out[20]: tensor(1.0000)                                                                                                                                                      
                                                                                                                                                                             
In [21]: v.det()                                                                                                                                                             
Out[21]: tensor(-1.0000)                                                                                                                                                     
                                                                                                                                                                             
In [22]: s                                                                                                                                                                   
Out[22]: tensor([1.9054, 0.3776, 0.0328])                                                                                                                                    
                                                                                                                                                                             
In [23]: x.det()                                                                                                                                                             
Out[23]: tensor(-0.0236)                                                                                                                                                     
                                                                                                                                                                             
In [24]: u.det() * v.det() * torch.prod(s)                                                                                                                                   
Out[24]: tensor(-0.0236)                         

In [30]: torch.autograd.gradcheck(lambda t: t.det(), x)                                                                                                              [3/1809]
/home/nik/git/pytorch/torch/autograd/gradcheck.py:273: UserWarning: Input #0 requires gradient and is not a double precision floating point or complex. This check will likel
y fail if all the inputs are not of double precision floating point or complex.                                                                                              
  f'Input #{idx} requires gradient and '                                                                                                                                     
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-30-b138b589287a> in <module>()
----> 1 torch.autograd.gradcheck(lambda t: t.det(), x)

/home/nik/git/pytorch/torch/autograd/gradcheck.py in gradcheck(func, inputs, eps, atol, rtol, raise_exception, check_sparse_nnz, nondet_tol, check_undefined_grad, check_grad_dtypes, check_batched_grad)
    623                 else:                 # R -> R, R -> C
    624                     if not torch.allclose(a, n, rtol, atol):
--> 625                         return fail_test(get_notallclose_msg(a, n, i, j))
    626
    627         if check_batched_grad:

/home/nik/git/pytorch/torch/autograd/gradcheck.py in fail_test(msg)
    575     def fail_test(msg):
    576         if raise_exception:
--> 577             raise RuntimeError(msg)
    578         return False
    579

RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[ 0.0726],
        [ 0.0382],
        [-0.0764],
        [-0.2254],
        [-0.3958],
        [ 0.3157],
        [ 0.1490],
        [ 0.3520],
        [-0.2645]])
analytical:tensor([[ 0.0732],
        [ 0.0382],
        [-0.0759],
        [-0.2201],
        [-0.3901],
        [ 0.3122],
        [ 0.1451],
        [ 0.3487],
        [-0.2613]])

```

Reply:
@kurtamohler , could you please add tests for inputs generated from the SVD exactly for cases `det(u) * det(v) == -1`?
If you have `det(x) == 1`, you can flit it to `-1` by randomly exchanging two columns/rows.

Reply:
@nikitaved, FWIW, your example above fails `gradcheck` because of floating point error (note the UserWarning), not because the formula is wrong. It passes if we cast it to double:

```python
>>> x = torch.tensor([[0.2931, 0.5179, 0.8540],
...     [0.6996, 0.3445, 0.8483],
...     [0.7505, 0.2611, 0.8554]], dtype=torch.float)
>>> torch.autograd.gradcheck(torch.det, x.requires_grad_(True))
UserWarning: Input #0 requires gradient and is not a double precision floating point or complex. This check will likely fail if all the inputs are not of double precision floating point or complex.
...
>>> torch.autograd.gradcheck(torch.det, x.clone().detach().to(torch.double).requires_grad_(True))
True
```

Also, this example isn't singular anyway, so it doesn't exercise `det_backward_singular_real`.

```python
>>> x.det()
tensor(-0.0236)
```

Reply:
Cool! Now it is going to be fixed :)